src/arkode/arkode.c

```diff
@@ -2995,6 +3001,7 @@ int arkCheckConvergence(ARKodeMem ark_mem, int* nflagPtr, int* ncfPtr)
   ARKodeHAdaptMem hadapt_mem;
 
   if (*nflagPtr == ARK_SUCCESS) { return (ARK_SUCCESS); }
+  if (*nflagPtr == ARK_RETRY_STEP) { return (ARK_RETRY_STEP); }
```

Comment:
- Are there comments documenting the purpose of this return value?
-- This is only an internal flag now. Just before every line this involves in the decision mechanism, we have added comments to justify why it should be as we implemented.

- When deciding to return `ARK_RETRY_STEP`, does the stepper check if fixed step sizes are being used? The current fixed step logic in `ARKodeEvolve` would accept this step rather than return an error.
-- This bug has been fixed. Thanks! That was a great catch. 

- Is this failure accounted for in the number of step failures?
--Since this return happens before stepper starts performing any computations, we do not count them as step failures.

- Is a new step failure counter needed since this is not a solver failure or error test failure?
-- This could be considered as part of the step size adaptivity as if it had not provided an appropriate step size. We immediately return to adjust it before doing any computations and throwing the step at the end. This has not much to do with the stepper even if it returns from the stepper, so we do not keep track of the number of such returns. 

- Is a new variable needed for the limit on the number of allowed failures of this type in a step attempt needed (analogous to `maxnef` and `maxncf`)?
-- These failures are not really failures if we consider this as part of the time step size adaptivity, I think the control mechanism for the time step adjustment will kick in, for example, when `h` get too small, etc. 

- Since this function covers more than convergence now but it should be renamed something like `arkCheckStepStatus` or `arkCheckStepReturn`
-- Actually this function checks the convergence as it did before. Only difference would be when there is no solver involves, this returns  without checking anything as it should be. There is not any other purpose of this function now.

Reply:
I believe that this topic is currently "resolved" through internal documentation.  Essentially, `ARK_RETRY_STEP` is allowable from a stepper's "take-step" routine under the conditions that:
(a) the stepper has updated `ark_mem->eta` (or perhaps something similar) so that the next step will run normally, and
(b) the stepper's "take-step" routine *cannot* return `ARK_RETRY_STEP` multiple times in a row.

As long as we trust ourselves (and potentially future contributors) to follow these rules, then no additional counters and "if" statements are needed.